clang/lib/Sema/SemaDecl.cpp

```diff
@@ -15978,6 +15988,24 @@ Decl *Sema::ActOnFinishFunctionBody(Decl *dcl, Stmt *Body,
       CheckCoroutineWrapper(FD);
   }
 
+  // Diagnose invalid SYCL kernel entry point function declarations.
+  if (FD && !FD->isInvalidDecl() && !FD->isTemplated() &&
+      FD->hasAttr<SYCLKernelEntryPointAttr>()) {
+    if (FD->isDeleted()) {
+      Diag(FD->getAttr<SYCLKernelEntryPointAttr>()->getLocation(),
+           diag::err_sycl_entry_point_invalid)
+          << /*deleted function*/ 2;
+    } else if (FD->isDefaulted()) {
+      Diag(FD->getAttr<SYCLKernelEntryPointAttr>()->getLocation(),
+           diag::err_sycl_entry_point_invalid)
+          << /*defaulted function*/ 3;
+    } else if (FSI->isCoroutine()) {
+      Diag(FD->getAttr<SYCLKernelEntryPointAttr>()->getLocation(),
+           diag::err_sycl_entry_point_invalid)
+          << /*coroutine*/ 7;
```

Comment:
When these get this high, an enum typically makes sense instead of comments (which get lost/etc).

Reply:
Can you point me to an example? I haven't been able to find one. It doesn't look like there is a way to declare an enum with the diagnostic definition. Without the ability to do so, I don't see how adding an intermediate enum somewhere else improves the situation.

Reply:
Yes, there isn't a convenient way of doing so in the diagnostics engine unfortunately.  I'd love for us to have that!  But typically we just define it elsewhere.  For large lists like this it can make it significantly more readable, and much less likely to not be updated because of a copy/paste error (which happens OFTEN).

See `AttributeArgumentNType` and `AttributeDeclKind` for some examples, but they are kind of sprinkled in quite a few places, Richard had me do some a few times IIRC for function multiversioning.  `Sema::checkTargetVersionAttr` has some more local versions of it.

Reply:
Thanks for pointing me to those examples. For both `AttributeArgumentNType` and `AttributeDeclKind`, the production of the diagnostic is more complicated. For the first case, the diagnostic is produced in a lambda expression and the `%select` index is passed as an argument. For the second case, the `%select` index is computed by mapping from another enumeration using a switch statement. An enum definitely seems warranted for those cases. Similarly, for `Sema::checkTargetVersionAttr`, there are multiple `%select` indices involved in each diagnostic and the local enumeration does help with readability.

For simple diagnostics, where the `%select` index doesn't require computation and where the same index values are not used in multiple places, I think indirection through an enumeration isn't helpful, so I haven't made a change. If you still think a change is warranted, perhaps we can get a second opinion.